Improving unit tests coverage

ezimuel opened the pull request:
This PR improves the unit tests coverage.

ezimuel committed:
WIP: improving unit tests

ezimuel committed:
Improve unit tests completed

weierophinney commented on src/ConfigTrait.php:
```diff
@@ -90,11 +90,16 @@ protected function getGrantsConfig(ContainerInterface $container) : array
-        if (empty($config['grants']) || ! is_array($config['grants'])) {
+        if (empty($config['grants'])) {
             throw new InvalidConfigException(
                 'The grant value is missing in config authentication'
```
This should be `"grants"`; otherwise, users might add a `grant` configuration value, and still get the exception. I would also likely add a note that the value must be an array.

ezimuel replied:
Ok, I changed in "grants" and added "must be an array" in the exception message.

weierophinney commented on src/ConfigTrait.php:
```diff
@@ -90,11 +90,16 @@ protected function getGrantsConfig(ContainerInterface $container) : array
         }
+        if (! is_array($config['grants'])) {
+            throw new InvalidConfigException(
+                'The grant must be an array value'
```
Same comment here.

ezimuel replied:
Changed.

weierophinney commented on test/ConfigTraitTest.php:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * @expectedException Zend\Expressive\Authentication\OAuth2\Exception\InvalidConfigException
+     */
+    public function testGetPrivateKeyNoConfig()
```
Update your test names to indicate what _behavior_ is being tested. In this case, the method would become something like `testGetPrivateKeyWhenNoConfigPresentWillResultInAnException`.

This practice makes understanding the _intention_ behind the test clearer when maintainers are trying to understand a failure later.

ezimuel replied:
Ok, I supposed was clear enough :)

weierophinney commented on test/ConfigTraitTest.php:
```diff
@@ -0,0 +1,154 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-authentication-oauth2 for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (https://www.zend.com)
```
Here and elsewhere you introduce new test classes: since the file was introduced this year, use 2018 as the copyright year.

weierophinney commented on test/Entity/AccessTokenEntityTest.php:
```diff
@@ -0,0 +1,24 @@
+use PHPUnit\Framework\TestCase;
+use Zend\Expressive\Authentication\OAuth2\Entity\AccessTokenEntity;
+
+class AccessTokenEntityTest extends TestCase
```
I don't think we need this test. If it's just testing that the constructor returns an instance of the class, that's something we know already.

If there's other behavior to test — e.g., default property values, property values based on constructor arguments, etc. — then tests make sense.

ezimuel replied:
If I remove the AccessTokenEntityTest we need to use `@codeCoverageIgnore` to omit this class from the code coverage. IMHO I think a test class like this, even if just checking for constructor, can be useful as a constrain for future changes. For instance, if someone changes the `AccessTokenEntityInterface` this test is able to intercept it.

weierophinney commented on test/Entity/AuthCodeEntityTest.php:
```diff
@@ -0,0 +1,24 @@
+
+class AuthCodeEntityTest extends TestCase
+{
+    public function testConstructor()
```
Same comment here as for the `AccessTokenTest` above.

weierophinney commented on test/Entity/ClientEntityTest.php:
```diff
@@ -0,0 +1,51 @@
+    {
+        $this->entity = new ClientEntity('foo', 'bar', 'http://localhost');
+    }
+    public function testConstructor()
```
This test method isn't necessary.

weierophinney commented on test/Entity/ClientEntityTest.php:
```diff
@@ -0,0 +1,51 @@
+{
+    public function setUp()
+    {
+        $this->entity = new ClientEntity('foo', 'bar', 'http://localhost');
```
I would write tests that ensure that the values passed to the constructor affected the state of the constructed entity.

weierophinney commented on test/Entity/RefreshTokenEntityTest.php:
```diff
@@ -0,0 +1,24 @@
+
+class RefreshTokenEntityTest extends TestCase
+{
+    public function testConstructor()
```
Same comment here as for the `AccessTokenTest`.

ezimuel replied:
I renamed to `testImplementsRefreshTokenEntityInterface` to test the interface implementation, not just the constructor.

weierophinney commented on test/Entity/RevokableTraitTest.php:
```diff
@@ -0,0 +1,27 @@
+
+class RevokableTraitTest extends TestCase
+{
+    public function testRevoked()
```
Break this into two tests, or use a test provider, to detail what happens with each of the boolean arguments.

ezimuel replied:
Done.

weierophinney commented on test/Entity/ScopeEntityTest.php:
```diff
@@ -0,0 +1,34 @@
+        $this->entity = new ScopeEntity();
+    }
+
+    public function testConstructor()
```
This method is unnecessary.

ezimuel replied:
I renamed into `testImplementsScopeEntityInterface`.

weierophinney commented on test/Entity/TimestampableTraitTest.php:
```diff
@@ -0,0 +1,43 @@
+
+namespace ZendTest\Expressive\Authentication\OAuth2\Entity;
+
+use DateTime;
```
Quick question: should we be using `DateTime`, or `DateTimeImmutable`? Which would make more sense for the domain?

Xerkus replied:
I would say accept `DateTimeInterface`, convert to `DateTimeImmutable` if mutable.

ezimuel replied:
I used `DateTimeInterface` as parameters of  `TimestampableTrait` and `DateTimeImmutable` in `TimestampableTrait::timestampOnCreate()` function.

weierophinney commented on test/Entity/UserEntityTest.php:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * @expectedException ArgumentCountError
+     */
+    public function testConstructorWithoutParam()
```
This test doesn't have any assertions. Assert against the default identifier.

ezimuel replied:
The assertion is in the phpdoc, I rename the function to `testConstructorWithoutParamWillResultInAnException`.

weierophinney reviewed (changes requested):
First off, great work - you jumped from 27% to 85% coverage!

There's still some room for improvement, though the improvements I suggest may not necessarily add coverage; they'll mainly make intent of tests more clear. In broad strokes:

- Testing constructors to ensure they create an instance of the class is mostly useless, _unless you also test for default property values_.
- Rename the various test methods to describe _the behavior being tested_. This will make the intent more clear, and may help you identify when you need to split a test into multiple tests and/or use a provider.

Xerkus commented:
I would suggest explicit class level `@covers` annotations with `forceCoversAnnotation="true"` phpunit option to avoid false/accidental coverage

ezimuel committed:
Resolved comments on PR #49

ezimuel commented:
@weierophinney I updated the PR including your feedback.

weierophinney committed:
Minor CS/consistency changes

- Commas after all array entries
- When splitting to multiple lines, move all arrow operations to their
  own lines when operating on an object

weierophinney committed:
Write tests for ClientEntity constructor

Writes three test methods that ensure the arguments passed to the
constructor affect the instance state.

weierophinney committed:
Fixes typo in method name of RevokableTraitTest

weierophinney committed:
Better test method name in ScopeEntityTest

s/testJsonSerialize/testEntityIsJsonSerializable/

weierophinney committed:
One line per object operation

weierophinney committed:
Updates copyright dates on all files touched by this patch

- New files: 2018
- Existing files: range adds "-2018"

weierophinney committed:
Adds php-nightly build to travis

So we can test against PHP 7.3.
